Use a Job as the Reconciler Thread for AbstractReconciler

merks opened the pull request:
Currently AbstractReconciler uses a raw thread for reconciling requests. This has the problem that if the job framework is suspended (e.g, during startup of the IDE) some code is still executed because the reconciler is called from the creation of the text viewer and starts immediately.

This replaces the bare thread with a job but retains current behaviour. Note that the job cannot be manually terminated, just as the thread continued to run indefinitely until explicitly terminated.

https://github.com/eclipse-platform/eclipse.platform.ui/pull/3018

merks commented:
@iloveeclipse 

I was able to reproduce the problem you described and with this change the job continues to run which avoids the described problem.  In addition, I see that the editor starts with error markers that are subsequently cleaned up...

laeubi reviewed (approved):
Thant for catching up with this, ignoring the cancel request seems like a reasonable approach here, maybe we even want to override `cancel()` of the job and make it a noop with a comment why it is done here that way.

merks commented:
> Thant for catching up with this, ignoring the cancel request seems like a reasonable approach here, maybe we even want to override `cancel()` of the job and make it a noop with a comment why it is done here that way.

For our safety and comfort, that's not possible.  😁 

https://github.com/eclipse-platform/eclipse.platform/blob/c3dfa7bc07048f1db682284b32a4011290b81c55/runtime/bundles/org.eclipse.core.jobs/src/org/eclipse/core/runtime/jobs/Job.java#L289-L291

iloveeclipse commented on bundles/org.eclipse.jface.text/src/org/eclipse/jface/text/reconciler/AbstractReconciler.java:
```diff
@@ -72,8 +79,8 @@ class BackgroundThread extends Thread {
-			setPriority(Thread.MIN_PRIORITY);
-			setDaemon(true);
+			setPriority(Job.DECORATE);
+			setSystem(true);
```
The job should provide a family so it can be queried from outside whether the reconciler of specific family is still running or not. Possible would be to have one family for a dedicated viewer for which this job is started, this would require the viewer to be passed as argument here. If that is too much, the minimal requirement would be that they should answer `true` to `job.belongsTo(AbstractReconciler.class)` (without introducing new API here).

See `org.eclipse.core.runtime.jobs.Job.belongsTo(Object)`

Background: in our tests (also in JDT debugger and I guess other places) we often check whether some jobs are still running after some operation, waiting that everything is "calm" before the next test step or on teardown.

Having here a family would allow to ignore jobs of that family without hard coding something like job.getClass().getName().equals ...

merks replied:
Sounds good.  Like this perhaps?
```
		@Override
		public boolean belongsTo(Object family) {
			return family == fViewer || AbstractReconciler.class.equals(family);
		}
```

iloveeclipse replied:
Yes, and you can use == instead of equals for class object.

merks replied:
Done.

iloveeclipse commented on bundles/org.eclipse.jface.text/src/org/eclipse/jface/text/reconciler/AbstractReconciler.java:
```diff
@@ -449,7 +471,7 @@ public void install(ITextViewer textViewer) {
 			if (fThread != null)
 				return;
-			fThread= new BackgroundThread(getClass().getName());
+			fThread= new BackgroundThread(getClass().getName() + " reconciler thread"); //$NON-NLS-1$
```
I believe we should better undo this change, just keep the class name as before, additional text it doesn't bring much and in JDT case makes the job name definitely not better (beside the fact that this is a "job" now):

![image](https://github.com/user-attachments/assets/b5689c63-d746-4952-a661-35ec74b99686)

merks replied:
Like this I believe, right?

![image](https://github.com/user-attachments/assets/388b0cac-9bdd-480f-b882-aee339a03ff4)

iloveeclipse replied:
Yes, just like it was before.

merks replied:
Done.

iloveeclipse reviewed (changes requested):
This looks better now, except two places.

I guess we might see some test hangs or slowdown because of that change for tests that open editors and do `waitForJobs()`, we will see...

Ideally if we submit that it would be **after** we have at least one "green" build in SDK.

merks commented:
@iloveeclipse 

All the failures seem related to an unrelated API filter problem:

![image](https://github.com/user-attachments/assets/0b0c7983-fb17-4919-afba-5700368f83b1)

iloveeclipse commented:
> @iloveeclipse
> 
> All the failures seem related to an unrelated API filter problem:
> 
> ![image](https://private-user-images.githubusercontent.com/208716/452276809-0b0c7983-fb17-4919-afba-5700368f83b1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDkyMDkwNjMsIm5iZiI6MTc0OTIwODc2MywicGF0aCI6Ii8yMDg3MTYvNDUyMjc2ODA5LTBiMGM3OTgzLWZiMTctNDkxOS1hZmJhLTU3MDAzNjhmODNiMS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNjA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDYwNlQxMTE5MjNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05Nzk5ZmFkNjVkYzI1YWI3ZjE2ZjE5YzRlNjAwYzNiNzYyYzAxYjExYjhhNzFhNDk5NTU3ZDUxMjk3MTEwYjUxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.TqSB-39cJQSwObpGeU30G5tQ4yP92vqK2pwFKwsJB4k)

@raghucssit : I guess it's same issue you mentioned?

merks commented:
@iloveeclipse 

I'll look at fixing that API issue too, but it's hard to juggle back and forth.  Also Ant current has errors in my workspace and Maven Publishing stopped working.  Too many fires to fight.  😱 

So if you're okay with the current content of this PR, maybe we could clear it out of the way?

iloveeclipse commented on bundles/org.eclipse.jface.text/src/org/eclipse/jface/text/reconciler/junk:
```diff
@@ -0,0 +1,665 @@
+/*******************************************************************************
```
@merks : looks like you've added "junk" file?

merks replied:
Oh dear.  What an idiot!

I will rebase and force push with the API warning thing fixed a little later.

Sorry for the stupid noise.

merks committed:
Use a Job as the Reconciler Thread for AbstractReconciler

Currently AbstractReconciler uses a raw thread for reconciling requests.
This has the problem that if the job framework is suspended (e.g, during
startup of the IDE) some code is still executed because the reconciler
is called from the creation of the text viewer and starts immediately.

This replaces the bare thread with a job but retains current
behaviour. Note that the job cannot be manually terminated, just as the
thread continued to run indefinitely until explicitly terminated.

https://github.com/eclipse-platform/eclipse.platform.ui/pull/3018

merks commented:
Hopefully it builds this time with no failures!

eclipse-platform-bot committed:
Version bump(s) for 4.37 stream

merks commented:
@iloveeclipse 

We even have a green build now.   😁

iloveeclipse reviewed (approved):
Ed, please check SDK test results on the next SDK build. I fear there could be some hangs.

merks commented:
I will look at the builds tomorrow.

merks commented:
FYI, I started this build:

https://ci.eclipse.org/releng/job/Builds/job/I-build-4.37/14/

iloveeclipse commented:
Probably related regression: https://github.com/eclipse-jdt/eclipse.jdt.ui/issues/2323
